jax/experimental/host_callback.py

```diff
@@ -1066,10 +1619,13 @@ def barrier_tap(dev_idx, _):
   with lock:
     cv.wait_for(lambda: num_at_large == 0)
   logging.vlog(2, f"barrier_wait[{logging_name}]: done")
-  if _outfeed_receiver.num_tap_exceptions > 0:
-    _outfeed_receiver.num_tap_exceptions = 0
+  if _outfeed_receiver.last_tap_exception is not None:
+    last_exception = _outfeed_receiver.last_tap_exception
+    _outfeed_receiver.last_tap_exception = None
     raise TapFunctionException(
-        "There were exceptions during id_tap processing.")
+        "There were exceptions during id_tap processing. "
+        f"Last one was: {last_exception}")
```

Comment:
I would suggest either re-raising the original error:
```python
    raise last_exception
```

Or maybe raising `TapFunctionException` _from_ the original error:
```python
    raise TapFunctionException(
        "There were exceptions during id_tap processing.") from last_exception
```
This ends up with an error message like:
```
(original traceback)
(original error)

The above exception was the direct cause of the following exception:

(barrier_tap traceback)
(TapFunctionException error)
```

Reply:
I have implemented the latter suggestion (`raise ... from last_exception`), and it works as you predicted only for `id_tap`. For errors during `call`, at least on CPU, the XLA runtime crashes (because we infeed to it a bad shape) and in that case the `from last_exception` does not seem to work.